python/cudf/cudf/core/dataframe.py

```diff
@@ -1555,13 +1438,30 @@ def insert(self, loc, column, value, forceindex=False):
                     -(num_cols + 1) * (num_cols > 0), num_cols * (num_cols > 0)
                 )
             )
-        self._cols[column] = self._prepare_series_for_add(
-            value, forceindex=forceindex, name=column
-        )
-        keys = list(self._cols.keys())
+
+        if is_scalar(value):
+            value = utils.scalar_broadcast_to(value, len(self))
+
+        if len(self) == 0:
+            if isinstance(value, (pd.Series, Series)):
+                self._index = as_index(value.index)
+            else:
+                self._index = RangeIndex(start=0, stop=len(value))
+                if num_cols != 0:
+                    for col_name in self._data:
+                        self._data[col_name] = column.column_empty_like(
+                            self._data[col_name],
+                            masked=True,
+                            newsize=len(value),
+                        )
+
+        value = column.as_column(value)
+
+        self._data[name] = value
+        keys = list(self._data.keys())
```

Comment:
Guessing you can't enumerate a dict_keys instance?

Reply:
You can enumerate an `iter(dict_keys)` instance as we discussed, but that results in a `RuntimeError: OrderedDict mutated during iteration`. It looks like our logic of using `move_to_end` during iteration [may need changing](https://bugs.python.org/issue24369).

[There are some approaches](https://stackoverflow.com/questions/38592779/how-to-implement-insert-for-ordereddict-in-python-3) for inserting into an `OrderedDict` that we might be able to adopt instead, but I think that's for a follow-up.